krm-functions/lib/kptfile/v1/kpt.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2023 Nephio.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptv1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"sigs.k8s.io/yaml"
+)
+
+type KptFile interface {
+	// Marshal serializes the value provided into a YAML document based on "sigs.k8s.io/yaml".
+	// The structure of the generated document will reflect the structure of the value itself.
+	Marshal() ([]byte, error)
+	// ParseKubeObject returns a fn sdk KubeObject; if something failed an error
+	// is returned
+	ParseKubeObject() (*fn.KubeObject, error)
+	// GetKptFile returns the Kptfile as a go struct
+	GetKptFile() *kptv1.KptFile
+	// SetConditions sets the conditions in the kptfile. It either updates the entry if it exists
+	// or appends the entry if it does not exist.
+	SetConditions(...kptv1.Condition)
+	// DeleteCondition deletes the condition equal to the conditionType if it exists
+	DeleteCondition(ct string)
+	// GetCondition returns the condition for the given ConditionType if it exists,
+	// otherwise returns nil
+	GetCondition(ct string) *kptv1.Condition
+	// GetConditions returns all the conditions in the kptfile. if not initialized it
+	// returns an emoty slice
+	GetConditions() []kptv1.Condition
+}
+
+// New creates a new parser interface
+// It expects a raw byte slice as input representing the serialized yaml file
+func New(b string) (KptFile, error) {
+	k := &kptv1.KptFile{}
+	if err := yaml.Unmarshal([]byte(b), k); err != nil {
```

Comment:
I'm guessing any comments in the Kptfile get lost here?

Reply:
Good Question. I did a quick test and I don't see comments getting lost when unmarshaling using this method.

Reply:
this is the outcome of the test I did.

from:

apiVersion: kpt.dev/v1
kind: Kptfile
# comment A
metadata:
  name: pkg-upf
  #commentB
  annotations:
    config.kubernetes.io/local-config: "true"
info:
  description: upf package example
pipeline: {}


apiVersion: kpt.dev/v1
kind: Kptfile
# comment A
metadata:
  name: pkg-upf
  #commentB
  annotations:
    config.kubernetes.io/local-config: "true"
info:
  description: upf package example
pipeline: {}
status:
  conditions:
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n3
    status: "False"
    type: ipam.nephio.org/v1alpha1.IPAllocation.n3
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n3
    status: "False"
    type: k8s.cni.cncf.io/v1.NetworkAttachmentDefinition.n3
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n4
    status: "False"
    type: ipam.nephio.org/v1alpha1.IPAllocation.n4
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n4
    status: "False"
    type: k8s.cni.cncf.io/v1.NetworkAttachmentDefinition.n4
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n6
    status: "False"
    type: ipam.nephio.org/v1alpha1.IPAllocation.n6
  - message: create new resource
    reason: req.nephio.org/v1alpha1.Interface.n6
    status: "False"
    type: k8s.cni.cncf.io/v1.NetworkAttachmentDefinition.n6